src/battle_ai_switch_items.c

```diff
@@ -899,37 +1080,764 @@ static u32 GetBestMonDmg(struct Pokemon *party, int firstId, int lastId, u8 inva
     int i, j;
     int dmg, bestDmg = 0;
     int bestMonId = PARTY_SIZE;
+    u32 aiMove;
 
     gMoveResultFlags = 0;
     // If we couldn't find the best mon in terms of typing, find the one that deals most damage.
     for (i = firstId; i < lastId; i++)
     {
         if (gBitTable[i] & invalidMons)
             continue;
-        if (IsAiPartyMonOHKOBy(opposingBattler, &party[i]))
+        InitializeSwitchinCandidate(&party[i]);
+        for (j = 0; j < MAX_MON_MOVES; j++)
+        {
+            if (aiMove != MOVE_NONE && gBattleMoves[aiMove].power != 0)
+            {
+                aiMove = GetMonData(&party[i], MON_DATA_MOVE1 + j);
+                dmg = AI_CalcPartyMonDamage(aiMove, gActiveBattler, opposingBattler, &switchinCandidate, TRUE);
+                if (bestDmg < dmg)
+                {
+                    bestDmg = dmg;
+                    bestMonId = i;
+                }
+            }
+        }
+    }
+
+    return bestMonId;
+}
+
+// Gets hazard damage
+static u32 GetSwitchinHazardsDamage(void)
+{
+    u8 defType1 = switchinCandidate.type1, defType2 = switchinCandidate.type2, tSpikesLayers;
+    u16 heldItemEffect = gItems[switchinCandidate.item].holdEffect;
+    u32 maxHP = switchinCandidate.maxHP, ability = switchinCandidate.ability, status = switchinCandidate.status1;
+    u32 spikesDamage = 0, tSpikesDamage = 0, hazardDamage = 0;
+    u32 hazardFlags = gSideStatuses[GetBattlerSide(gActiveBattler)] & (SIDE_STATUS_SPIKES | SIDE_STATUS_STEALTH_ROCK | SIDE_STATUS_STICKY_WEB | SIDE_STATUS_TOXIC_SPIKES);
+
+    // Check ways mon might avoid all hazards
+    if (ability != ABILITY_MAGIC_GUARD || (heldItemEffect == HOLD_EFFECT_HEAVY_DUTY_BOOTS &&
+        !((gFieldStatuses & STATUS_FIELD_MAGIC_ROOM) || ability == ABILITY_KLUTZ)))
+    {
+        // Stealth Rock
+        if ((hazardFlags & SIDE_STATUS_STEALTH_ROCK) && heldItemEffect != HOLD_EFFECT_HEAVY_DUTY_BOOTS)
+            hazardDamage += GetStealthHazardDamageByTypesAndHP(gBattleMoves[MOVE_STEALTH_ROCK].type, defType1, defType2, switchinCandidate.hp);
+        // Spikes
+        if ((hazardFlags & SIDE_STATUS_SPIKES) && ((defType1 != TYPE_FLYING && defType2 != TYPE_FLYING
+            && ability != ABILITY_LEVITATE && heldItemEffect != HOLD_EFFECT_AIR_BALLOON)
+            || heldItemEffect == HOLD_EFFECT_IRON_BALL || gFieldStatuses & STATUS_FIELD_GRAVITY))
+        {
+            spikesDamage = maxHP / ((5 - gSideTimers[GetBattlerSide(gActiveBattler)].spikesAmount) * 2);
+            if (spikesDamage == 0)
+                spikesDamage = 1;
+            hazardDamage += spikesDamage;
+        }
+        // Toxic Spikes
+        if ((hazardFlags & SIDE_STATUS_TOXIC_SPIKES) && ((defType1 != TYPE_POISON && defType2 != TYPE_POISON
```

Comment:
Left out this change as it required 
```
struct BattlePokemon *savedBattleMons = AllocSaveBattleMons();
gBattleMons[battler] = switchinCandidate.battlemon
[Function]
FreeRestoreBattleMons(savedBattleMons);
```
which added extra time over just using embedding the conditional here. I did update the conditional to consider everything in CanBePoisoned though. I'd rather some duplicate code to run faster in this case, given that extending runtime makes the player wait longer in all these functions.

Reply:
I see, in this case you can remove the poison type check and use `CanPoisonType` instead. For the flying type check use IS_BATTLER_OF_TYPE

Reply:
These both do the same thing though where they accept a battlerID rather than a BattlePokemon struct, and to assign that battlerID to our struct we'd still have to go through the steps I mentioned, no?

I have added Steel types from the CanPoisonType check though, somehow forgot about those :)

Reply:
Dang, you're right :D I guess these functions should be somehow merged to avoid code duplication. For now it can stay the way it is, I'd ask you to just add a comment with something like // TODO: don't duplicate code with CanBePoisoned